overlays/haskell.nix

```diff
@@ -219,24 +219,49 @@ self: super: {
             '';
 
         # Given a list of repos:
-        # [ { name = ...; url = ...; rev = ...; sha256 = ... } ]
+        # [ { name = ...; url = ...; rev = ...; ref = ...; sha256 = ...; is-private = ...; } ]
         # produce a cache file that can be used for
         # stack-to-nix or plan-to-nix to prevent them
         # from needing network access.
-        mkCacheFile = repos: let
-            fetchRepo = repo: (self.buildPackages.pkgs.fetchgit {
-                url = repo.url;
-                rev = repo.rev;
-                sha256 = repo.sha256;
-            }) + (if repo.subdir or "." == "." then "" else "/" + repo.subdir);
-
-            f = { name, url, rev, subdir ? ".", sha256, cabal-file ? "${name}.cabal" }@repo:
-                let nix-expr = self.buildPackages.haskell-nix.callCabalToNix { src = (fetchRepo repo); inherit name cabal-file; };
-                in "${url} ${rev} ${subdir} ${sha256} ${name} ${nix-expr}";
+        mkCacheFile = repos:
+          let
+            # Fetch the entire repo, using either pkgs.fetchgit or
+            # builtins.fetchGit depending on whether the repo is private.
+            fetchEntireRepo = repo:
+              if repo.is-private or false
+                then
+                  # It doesn't make sense to specify sha256 on a private repo
+                  # because it is not used by buitins.fetchGit.
+                  assert !(repo ? sha256);
+                  builtins.fetchGit {
+                    url = repo.url;
+                    rev = repo.rev;
+                  } // self.buildPackages.lib.optionalAttrs (repo ? ref) {
+                    ref = repo.ref;
+                  }
+                else
+                  # Non-private repos must have sha256 set.
+                  assert repo ? sha256;
+                  # pkgs.fetchgit doesn't have any way of fetching from a given
+                  # ref.
+                  assert !(repo ? ref);
```

Comment:
It might be nicer to have these assertions return an understandable error message, but for now they should be fine.